### PR TITLE
Add helm chart for nats connector

### PR DESCRIPTION
**Pull request by viveksyngh**
Signed-off-by: Vivek Singh <vivekkmr45@yahoo.in>

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: chart/nats-connector/values.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+replicas: 1
+image: openfaas/nats-connector:0.2.2
```
Comment (Waterdrips) [inline chart/nats-connector/values.yaml:2]: helm convention is to use camel case, and also include a comment with each/most values giving some indication of what they are used for in the chart

File: chart/nats-connector/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+Specify each parameter using the `--set key=value[,key=value]` argument to `helm install`.
+See values.yaml for detailed configuration.
+
+## Removing the nats-connector
+
+All control plane components can be cleaned up with helm:
+
+```sh
+$ helm delete --purge nats-connector
```
Comment (developer-guy) [inline chart/nats-connector/README.md:48]: I think we should say users which are actually using helm 2 you should use "helm delete --purge nats-connector" but on the other hand for helm 3 users we should say you should use "helm uninstall nats-connector".

File: chart/nats-connector/templates/deployment.yml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+          volumeMounts:
+            - name: auth
+              readOnly: true
+              mountPath: "/var/secrets/"
+          {{- end }}
+      {{- if .Values.basic_auth }}
+      volumes:
+        - name: auth
+          secret:
+            secretName: basic-auth
```
Comment (developer-guy) [inline chart/nats-connector/templates/deployment.yml:77]: I can't see any secret.yaml file which name is "basic-auth", should we need to add it?

File: chart/nats-connector/templates/deployment.yml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+              value: {{ .Values.gateway_url | quote }}
+            - name: topics
+              value: {{ .Values.topics | quote }}
+            - name: print_response
+              value: {{ .Values.print_response | quote }}
+            - name: print_response_body
+              value: {{ .Values.print_response_body | quote }}
+            {{- if .Values.basic_auth }}
+            - name: basic_auth
+              value: "true"
```
Comment (developer-guy) [inline chart/nats-connector/templates/deployment.yml:53]: i think this one's value  also needs to come from values.yaml : {{ .Values.basic_auth }}

File: chart/nats-connector/templates/deployment.yml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            - name: secret_mount_path
+              value: "/var/secrets/"
```
Comment (developer-guy) [inline chart/nats-connector/templates/deployment.yml:55]: should we move that env's value to values.yaml ?

File: chart/nats-connector/values.yaml (hunk 1 above)
Comment (viveksyngh) [reply chart/nats-connector/values.yaml:2]: values is kept like that to make it compatible with existing connectors. May be we can create a new issue to update the values across all connectors chart.

cc @alexellis

Comment (viveksyngh) [reply chart/nats-connector/values.yaml:2]: I will add comments for each values.

File: chart/nats-connector/templates/deployment.yml (hunk 3 above)
Comment (viveksyngh) [reply chart/nats-connector/templates/deployment.yml:77]: These secrets are created while OpenFaas deployment that's why there is no secret.yaml file.

**Comment by viveksyngh** [comment]
Tested on k3d cluster by deploying nats connector with local chart.

Add and install nats connector with helm chart
```
 helm repo add nats-connector  https://openfaas.github.io/faas-netes
"nats-connector" has been added to your repositories

 helm install nats-connector nats-connector --namespace openfaas 
NAME: nats-connector
LAST DEPLOYED: Mon May  4 23:26:46 2020
NAMESPACE: openfaas
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
Thanks for installing nats-connector. Please follow the instructions below to get you started.

You can watch the Connector logs to see it invoke your functions:

$ kubectl logs -n openfaas deploy/nats-connector -f
```

Deploy `recieve-message` function
```
 faas-cli deploy --name receive-message --image openfaas/receive-message:latest --fprocess='./handler' --annotation topic="nats-test"
```

Deploy `publish-message` function
```
faas-cli deploy --name publish-message --image openfaas/publish-message:latest --fprocess='./handler' --env nats_url=nats://nats.openfaas:4222
```

Invoke `publish-message` function
```
 faas-cli invoke publish-message <<< "test message"
```

Log from connector
```
kubectl logs -n openfaas deploy/nats-connector -f
==============================================================================
NATS Connector for OpenFaaS

Gateway URL: http://gateway.openfaas:8080
NATS URL: nats://nats:4222
Topics: 30s
Upstream timeout: [nats-test]
Topic-map rebuild interval: 5s
Async invocation: "false"
==============================================================================

2020/05/04 17:56:57 Configured topics: [nats-test]
2020/05/04 17:56:57 Binding to topic: "nats-test"
2020/05/04 17:56:57 Subscription: nats-test ready
2020/05/04 18:06:14 Topic: nats-test, message: "test message"
2020/05/04 18:06:14 Invoke function: receive-message.openfaas-fn
2020/05/04 18:06:14 connector-sdk got result: [200] nats-test => receive-message.openfaas-fn (28) bytes
```

Log of receive-message
```
aas-cli logs receive-message
WARNING! Communication is not secure, please consider using HTTPS. Letsencrypt.org offers free SSL/TLS certificates.
2020-05-04T18:05:57Z Forking - ./handler []
2020-05-04T18:05:57Z 2020/05/04 18:05:57 Started logging stderr from function.
2020-05-04T18:05:57Z 2020/05/04 18:05:57 Started logging stdout from function.
2020-05-04T18:05:57Z 2020/05/04 18:05:57 OperationalMode: http
2020-05-04T18:05:57Z 2020/05/04 18:05:57 Timeouts: read: 10s, write: 10s hard: 10s.
2020-05-04T18:05:57Z 2020/05/04 18:05:57 Listening on port: 8080
2020-05-04T18:05:57Z 2020/05/04 18:05:57 Writing lock-file to: /tmp/.lock
2020-05-04T18:05:57Z 2020/05/04 18:05:57 Metrics listening on port: 8081
2020-05-04T18:06:14Z 2020/05/04 18:06:14 stderr: 2020/05/04 18:06:14 received "test message"
2020-05-04T18:06:14Z 2020/05/04 18:06:14 POST / - 200 OK - ContentLength: 28
```

File: chart/nats-connector/README.md (hunk 2 above)
Comment (viveksyngh) [reply chart/nats-connector/README.md:53]: Added

File: chart/nats-connector/templates/deployment.yml (hunk 4 above)
Comment (viveksyngh) [reply chart/nats-connector/templates/deployment.yml]: Updated

**Comment by alexellis** [comment]
```
 helm repo add nats-connector  https://openfaas.github.io/faas-netes
"nats-connector" has been added to your repositories

 helm install nats-connector nats-connector --namespace openfaas 
```

I'm confused how this worked, since the chart isn't published yet?

**Comment by alexellis** [comment]
One additional patch is required for these helm chart additions:

```
diff --git a/Makefile b/Makefile
index 4551b2a..c3b91d9 100644
--- a/Makefile
+++ b/Makefile
@@ -28,7 +28,7 @@ install-armhf: namespaces
        kubectl apply -f yaml_armhf/
 
 charts:
-       cd chart && helm package openfaas/ && helm package kafka-connector/ && helm package cron-connector/
+       cd chart && helm package openfaas/ && helm package kafka-connector/ && helm package cron-connector/ && helm package nats-connector/
        mv chart/*.tgz docs/
        helm repo index docs --url https://openfaas.github.io/faas-netes/ --merge ./docs/index.yaml
        ./contrib/create-static-manifest.sh
```

This is so that a maintainer can run `make charts` and publish new files from the `docs` folder.

**Comment by alexellis** [comment]
I'll make this change today, but you'll see the new lines added like so:

```
space-mini:faas-netes alex$ export PATH=$PATH:~/.arkade/bin/
space-mini:faas-netes alex$ make charts
cd chart && helm package openfaas/ && helm package kafka-connector/ && helm package cron-connector/ && helm package nats-connector/
Successfully packaged chart and saved it to: /Users/alex/go/src/github.com/openfaas/faas-netes/chart/openfaas-5.8.1.tgz
Successfully packaged chart and saved it to: /Users/alex/go/src/github.com/openfaas/faas-netes/chart/kafka-connector-0.4.0.tgz
Successfully packaged chart and saved it to: /Users/alex/go/src/github.com/openfaas/faas-netes/chart/cron-connector-0.3.1.tgz
Successfully packaged chart and saved it to: /Users/alex/go/src/github.com/openfaas/faas-netes/chart/nats-connector-0.1.0.tgz
mv chart/*.tgz docs/
helm repo index docs --url https://openfaas.github.io/faas-netes/ --merge ./docs/index.yaml
```

**Comment by viveksyngh** [comment]
> I'm confused how this worked, since the chart isn't published yet?

It worked because it picked chart information from the `nats-connector` directory which was in the path to current directory.

**Pull request merged**